test/successful-directories.test.js

```diff
@@ -31,447 +29,553 @@ afterAll(() => {
-              'a/b/c/d/e/f/.foorc',
-              'a/b/c/d/e/f/foo.config.js',
-              'a/b/c/d/e/package.json',
-            ]);
-
-            expect(result).toEqual({
-              config: { found: true },
-              filepath: absolutePath('a/b/c/d/e/package.json'),
-            });
-          },
-        ]);
-      }
-    );
-
-    testSyncAndAsync('finds JS file in first searched dir', sync => () => {
-      function readFile(searchPath) {
-        switch (searchPath) {
-          case absolutePath('a/b/c/d/e/f/package.json'):
-          case absolutePath('a/b/c/d/e/f/.foorc'):
-          case absolutePath('a/b/c/d/e/package.json'):
-          case absolutePath('a/b/c/d/e/.foorc'):
-          case absolutePath('a/b/c/d/e/foo.config.js'):
-            throw { code: 'ENOENT' };
-          case absolutePath('a/b/c/d/e/f/foo.config.js'):
-            return 'module.exports = { found: true };';
-          default:
-            throw new Error(`irrelevant path ${searchPath}`);
-        }
-      }
-      const readFileMock = mockReadFile(sync, readFile);
-      const startDir = absolutePath('a/b/c/d/e/f');
-
-      expect.hasAssertions();
-      return testFuncsRunner(sync, loadConfig(sync, startDir), [
-        result => {
-          util.assertSearchSequence(readFileMock, [
-            'a/b/c/d/e/f/package.json',
-            'a/b/c/d/e/f/.foorc',
-            'a/b/c/d/e/f/foo.config.js',
-          ]);
-
-          expect(result).toEqual({
-            config: { found: true },
-            filepath: absolutePath('a/b/c/d/e/f/foo.config.js'),
-          });
-        },
-      ]);
+describe('finds rc file in third searched dir, with a package.json lacking prop', () => {
+  const startDir = absolutePath('a/b/c/d/e/f');
+  const readFile = searchPath => {
+    switch (searchPath) {
+      case absolutePath('a/b/c/d/e/f/package.json'):
+      case absolutePath('a/b/c/d/e/f/.foorc'):
+      case absolutePath('a/b/c/d/e/f/foo.config.js'):
+      case absolutePath('a/b/c/d/e/package.json'):
+      case absolutePath('a/b/c/d/e/.foorc'):
+      case absolutePath('a/b/c/d/e/foo.config.js'):
+        throw { code: 'ENOENT' };
+      case absolutePath('a/b/c/d/package.json'):
+        return '{ "false": "hope" }';
+      case absolutePath('a/b/c/d/.foorc'):
+        return '{ "found": true }';
+      default:
+        throw new Error(`irrelevant path ${searchPath}`);
+    }
+  };
+
+  const checkResult = (readFileMock, result) => {
+    util.assertSearchSequence(readFileMock, [
+      'a/b/c/d/e/f/package.json',
+      'a/b/c/d/e/f/.foorc',
+      'a/b/c/d/e/f/foo.config.js',
+      'a/b/c/d/e/package.json',
+      'a/b/c/d/e/.foorc',
+      'a/b/c/d/e/foo.config.js',
+      'a/b/c/d/package.json',
+      'a/b/c/d/.foorc',
+    ]);
+
+    expect(result).toEqual({
+      config: { found: true },
+      filepath: absolutePath('a/b/c/d/.foorc'),
     });
+  };
+
+  test('async', () => {
+    const readFileMock = mockReadFile(false, readFile);
```

Comment:
Is there a point to instead just calling ``jest.spyOn`` directly here? Using ``jest.spyOn`` directly should simplify this.

```js
const readFileMock = jest.spyOn(fsMock, 'readFile').mockImplementation(readFile)
```

Reply:
Maybe a relic of pre-Jest days. I'd be fine with making this less indirect by using `jest.spyOn` directory 👍

Reply:
The helper function allowed to use the same `readFile` implementation for both sync and async version(which involves callbacks).